[GSoC][implementions] Provide basic structure for the implementations directory

Polaris000 opened the pull request:
This patch adds the basic structure for the latest reference implementations. Implementations will be added from the next pull request. 
Closes #154.

Polaris000 commented:
Please have a look at @jgbarah!

jgbarah commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+        else:
+            self.until = utils.get_date(self.df, "until")
+
+    def _clean_commit(self, line):
```
This would be similar to the `flatten_commit` that I mention above.

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+        """
+        Initilizes self.df, the dataframe with one commit per row.
+
+        :param data_list: A list of dictionaries.
```
Will do.

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+        Initilizes self.df, the dataframe with one commit per row.
+
+        :param data_list: A list of dictionaries.
+            Each element a line from the JSON file
```
Will make the required changes!

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+
+        :param data_list: A list of dictionaries.
+            Each element a line from the JSON file
+        :param date_range: A tuple which represents the period of interest
```
OK!

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+
+        clean_data_list = list()
+        self.since = date_range[0]
+        self.until = date_range[1]
```
Oops! That should n't be there : will remove that.

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+        :param data_list: A list of dictionaries.
+            Each element a line from the JSON file
+        :param date_range: A tuple which represents the period of interest
+        :param src_code_obj: An object of SourceCode.
```
For now, I'll make the changes you suggest. We'll make changes later on if required.

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+
+        super().__init__(data_list)
+
+        clean_data_list = list()
```
In my opinion,  it is a better name for the list of cleaned commit dictionaries. `items` may refer to even the uncleaned list of commits, especially when we move on to the non-pandas version: which would have an `unclean_list` as well as `clean_list`

Is there a particular reason you favor `items`?

Polaris000 commented on implementations/scripts/Commit.py:
```diff
@@ -0,0 +1,82 @@
+        self.since = date_range[0]
+        self.until = date_range[1]
+
+        for line in self.raw_df.iterrows():
```
#### What I have done
- Metric.py:
    + Gets a list of raw dictionaries from Perceval/json file
    + Flattens it using the flatten method in the same class
- Commit.py:
    + Based on since, until / sourcecode, excludes commits
    + Picks out certain commit attributes to create a final clean df
- CodeChanges.py
    + Computes metric / timeseries metric over the df after Commit has worked with it.

> In Commit (or maybe Code_Change) class, you would define flatten_item, as a function that given an item from Perceval, with a commit, produces a flat item (that, is, an flat dictionary) if conditions apply (eg, if date is as it should be, if it is source_code, etc.), and None if not.

But isn't the data flattened in the Metric class?
Are you suggesting to use a child class function (`flatten_item` defined in Commit.py or CodeChanges.py) in the parent class (Metric.py) and do all the flattening as well as cleaning at once in Metric.py?

Can you please elaborate a little @jgbarah ? Thanks. 
@aswanipranjal, if possible, could you give your views on this? Thanks :)

aswanipranjal replied:
@Polaris000, sorry, I am a little late in the discussion.
> But isn't the data flattened in the Metric class?

Correct. The implementations of the `flatten` method will be in the child classes (Commit, Issue).

> Are you suggesting to use a child class function (flatten_item defined in Commit.py or CodeChanges.py) in the parent class (Metric.py) and do all the flattening as well as cleaning at once in Metric.py?

Correct. Right now the current implementation which has been merged in the repo, in the Metric class is:
```
    def __init__(self, items):
        flat_items = self._flatten_data(items)
        self.raw_df = pd.DataFrame(flat_items)
```

The `_flatten_data` function should be implemented in the `Commit` class, which is what you've done and is the correct way.


What I think @jgbarah and my concern is that the following code in the `__init__` method of the `Commit` class:
```
self.issrccode_obj = issrccode_obj
(self.since, self.until) = date_range
clean_items = list()
for line in items:
    commit = self._flatten_data(line)
    if commit is not None:
        clean_items.append(commit)
self.df = pd.DataFrame(clean_items)
if self.since is None:
    self.since = utils.get_date(self.df, 'since')
if self.until is None:
        self.until = utils.get_date(self.df, 'until')
```

will be common to all the other data sources as well (Issue, PullRequest) and hence can be moved to the `Metric` class's `__init__` method.

This way, we have common code to initialize the classes and specific implementations to `flatten` or `clean` the data according to the data sources.

Is it more clear now @Polaris000?

Polaris000 replied:
Thanks @aswanipranjal! My latest idea (#175) involves moving the population of `since` and `until` to `compute_timeseries` method. This way, the __init__method would have nothing out of place. The Metric(root) class is not implemented anyway.

Polaris000 commented:
@jgbarah Thanks for the prompt review! I had a look and left a few comments where I was a little unsure. I'll combine the pull requests and create a new one. I thought they would be easier to review if separated.

Polaris000 commented:
I have added combined #160 and #161 into #162 and added commits there. Please have a look . Before that, please have a look at my comment here (just above this one). Thanks :)
